aiohttp/web_request.py

```diff
@@ -510,8 +510,26 @@ def text(self):
         return bytes_body.decode(encoding)
 
     @asyncio.coroutine
-    def json(self, *, loads=json.loads):
-        """Return BODY as JSON."""
+    def json(self, *, loads=json.loads, allowed_types=None, err_handler=None):
+        """ Return BODY as JSON. If the `Content-Type` is not in allowed_types,
+        then the request is rejected.
+        You can also handle it manually by passing the `err_handler` parameter.
+        """
+        wrong_format_flag = False
+        if isinstance(allowed_types, str):
+            wrong_format_flag = allowed_types != self.content_type
+        elif isinstance(allowed_types, list):
+            wrong_format_flag = self.content_type not in allowed_types
+        elif allowed_types is None:  # do not check the content_type
+            wrong_format_flag = False
+
+        # if fail to pass the check
+        if wrong_format_flag:
+            if err_handler is None:
+                raise HTTPNotAcceptable()
+            elif callable(err_handler):
+                yield from err_handler()
```

Comment:
As we discussed above, there's no need to give callbacks. Instead, only the 406 error is raised.